### PR TITLE
16.0 web editor fix pretty print dmo

### PR DESCRIPTION
**Pull request by dmo-odoo**

### CODE REVIEW

**Comment by dmo-odoo** [comment]
@xmo-odoo  I would like your opinion on the first commit. Basically, forever since https://github.com/odoo/odoo/commit/6b857b6eeb59137a71385f98c82c440ac82cd45d we have this case that breaks in website:
```html
<p><span>a</span><span>b</span></p>
```
Saving it results in:
```html
<p>
    <span>a</span>
    <span>b</span>
</p>
```
Which, when re-rendered back in the DOM renders equivalent to:
```html
<p><span>a</span> <span>b</span></p>
```
Note the space between "a" and "b". That is because etree will pretty print nodes with indentations as long as they do not have text content, and that indentation is collapsed into a single visible space by the browser when inserted in the DOM. I used span nodes in this example but the same applies to any inline node really. This is very easily reproduced in website on any version:
- Drop a Text snippet.
- Replace all the content of the snippet by "ab".
- Put "a" in bold and "b" in italic.
- Save.
- Notice that the saved version is now "**a** _b_" instead "**a**_b_"

Consider this:
```bash
>>> etree.tostring(html.fromstring('<p><span>a</span><span>b</span></p>'), pretty_print=True)​
​b'<p>\n    <span>a</span>\n    <span>b</span>\n</p>\n'​
```
Which is incorrect, while this:
>>> etree.tostring(html.fromstring('<p><span>a</span><span>b</span>c</p>'), pretty_print=True)​
​b'<p><span>a</span><span>b</span>c</p>\n'​
Is correct.

The proposed fix is a rather heavy hack that leverages this behavior by inserting one of the few unicode control characters that etree considers to be actual content, and therefore prevents pretty printing for this node. The control character is then removed by the server to avoid polluting the actual views.

An alternative to this would be to disable pretty printing altogether, since the original commit seem to have introduced it as a fix for an old version of the editor rather than for the intrinsic qualities of having pretty printed views, but I suppose people have realized the value of pretty printed views since then and they would probably be sad to see it go now.

Another alternative would just be to tradeoff the case, as it rarely happens in practice since you need to change the font style of contiguous characters without anything in between, no space, no delimiter, or anything. However, I would argue against it because there is actually a valid use case for which the user can fall into this trap, and that is exactly what the original user who encountered this bug was doing (reproducible in any version):
- Drop a Text snippet.
- Replace all the content of the snippet by "Our products are very reliable.".
- You want this phrase to be very visible so you select the entire sentence and apply to it a vibrant color from the color theme of your website.
- You want "reliable" to be even more visible so you apply a red color to it.
- Save.
- Notice that the saved version now displays "reliable ." instead of "reliable.".

I'm very open to suggestion if you see other possible alternatives on how to fix this bug. I already tried passing `method='html'` to `etree.tostring` but then it doesn't pretty print anything anymore as it knows it can create that sort of issues. They basically don't support pretty printing html as it seems, they only support it for xml, hence why we are running into that sort of issues.

Another alternative would be to implement the pretty printing on client side and have a flag of some sort to disable the pretty printing server side but that seems overkill IMO considering this bug has been going unnoticed since 8.0, indicating that encountering it in practice is actually pretty rare. This is especially true considering there is basically no user way to recover from this because once your HTML has received indentation, you're screwed. Your only option is to delete the snippet and re-create the same one, except this time you'll need to remember to add some unformated invisible characters like spaces at the end of the previously-affected content. So this case is very rarely encountered in practice, otherwise users would have likely complained about it by now.

TL;DR: Bug is rare but has no easy work around. Proposed fix is ugly. Any idea ?

File: addons/website/static/src/js/editor/wysiwyg.js [hunk 1]
```diff
@@ -114,6 +116,40 @@ const WebsiteWysiwyg = Wysiwyg.extend({
...
+            if (
+                children.length > 1 &&
+                !hasNonEmptyTextNode &&
+                children.every(child => !isBlock(child))
+            ) {
+                // Add a DEL control code as last sibling to trick the pretty
+                // printer into thinking there is actual text and therefore
+                // avoid pretty printing that particular element. The DEL
+                // control code will be removed by the server afterwards.
+                element.appendChild(document.createTextNode('\u007F'));
```
Comment (dmo-odoo) [inline addons/website/static/src/js/editor/wysiwyg.js:149]: `cleanForSave` ? More like `dirtyForSave`, amirite ? :trollface:

**Comment by dmo-odoo** [comment]
Right, the fact that everything that originally comes from XML is already formatted was not on my mind. Indeed with a little luck it might mean that only the actual content written by the user would not be formatted. I might try that next week when I come back from holidays to see if the new editor also removes existing indentation or not. That might be a side-effect of the browser or something, in which case we'd be screwed anyway.

However, even if pre-existing formatting is kept, the approach would still quickly result in some unsavory html like:
```html
<p>first paragraph</p><p>second paragraph</p><ul><li>list item 1</li><list item 2</li></ul>
```
Which is purely user-generated content and would therefore be on a single line.

**Comment by dmo-odoo** [comment]
Closing this in favor of https://github.com/odoo/odoo/pull/113150 after @antonylesuisse agreed to remove the pretty printing of `ir.ui.view`.
